lightning-invoice/src/utils.rs

```diff
@@ -218,6 +198,85 @@ where
 	}
 }
 
+/// Utility to filter channels for an invoice, and return the corresponding route hints to include
+/// in the invoice.
+///
+/// The filtering ensures that the RouteHints returned only include the highest inbound capacity
+/// channel per counterparty node for which channels are present. Channels with a lower inbound
+/// capacity than the invoice amount aren't included in the hints. If any of the channels are
+/// private, the function returns no RouteHints, and the sender will need to look at the public
+/// channels to find the path instead.
+///
+/// Input:
+/// `channels`: The channels to filter.
+/// `min_inbound_capacity_required`: Defines the lowest inbound capacity a channel must have to not
+/// be filtered out.
+///
+/// Result:
+/// `Vec<RouteHint>`: The filtered RouteHints
+/// `bool`: True if any of the inputted channels are private, else false
+fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_required: Option<u64>) -> (Vec<RouteHint>, bool){
+	let mut filtered_channels: HashMap<PublicKey, (u64, RouteHint)> = HashMap::new();
+	let min_capacity_required = match min_inbound_capacity_required {
+		Some(amt) => amt,
+		None => 0,
+	};
+	let route_hint_from_channel = |channel: ChannelDetails, short_channel_id: u64, forwarding_info: CounterpartyForwardingInfo| {
+        RouteHint(vec![RouteHintHop {
+            src_node_id: channel.counterparty.node_id,
+            short_channel_id,
+            fees: RoutingFees {
+                base_msat: forwarding_info.fee_base_msat,
+                proportional_millionths: forwarding_info.fee_proportional_millionths,
+            },
+            cltv_expiry_delta: forwarding_info.cltv_expiry_delta,
+            htlc_minimum_msat: None,
+            htlc_maximum_msat: None,}])
+    };
+	for channel in channels {
+		if !channel.is_public {
+			// If any of the channels are private, return no hints and let the sender look at the
+			// public channels instead.
+			let has_private_channels = true;
+			return (vec![], has_private_channels);
```

Comment:
Rather than returning a `bool`, can the caller simply check if the returned `Vec` is empty?

Reply:
That depends on on wether we would like the invoice to include single hop hint referencing a phantom payment-receiving node that only has public channels which inbound capacity are all below the invoice amt or not. 

Currently the code will not include a hint to those nodes. But If we're ok with including a hint to such nodes even though there are no channels that can be used to route to them (at least at the point of the invoice creation), that would certainly work!

That would definitely make the code cleaner :)

Reply:
Ah, that's an excellent point. I'd imagine we could have this method return `Result<Vec<RouteHint, ()>>` (or possible using a custom error type instead of `()`) and let the caller determine what to do if no channels have sufficient inbound capacity.

Reply:
True! Ok I'll make the function return a `Result` type instead.